File: ios/Modules/Sources/Reactions/impl/InMemoryReactions.swift
import Foundation

public final class InMemoryReactions: Reactions, @unchecked Sendable {
    private var counts: [Int64: [String: Int]] = [:]
    private var mine: [Int64: Set<String>] = [:]
    private let lock = NSLock()

    public init() {}

    public func reactions(for messageId: Int64) -> [Reaction] {
        lock.lock(); defer { lock.unlock() }
        let mineSet = mine[messageId] ?? []
        return (counts[messageId] ?? [:])
            .map { Reaction(messageId: messageId, emoji: $0.key, count: $0.value, reactedByMe: mineSet.contains($0.key)) }
            .sorted { $0.count > $1.count }
    }

    public func apply(messageId: Int64, emoji: String, delta: Int, byMe: Bool) {
        lock.lock(); defer { lock.unlock() }
        var m = counts[messageId] ?? [:]
        m[emoji, default: 0] += delta
        if m[emoji, default: 0] <= 0 { m.removeValue(forKey: emoji) }
        counts[messageId] = m
        if byMe {
            var mset = mine[messageId] ?? []
            if delta > 0 { mset.insert(emoji) } else { mset.remove(emoji) }
            mine[messageId] = mset
        }
    }
}

File: android/app/src/main/java/sol/vortexx/android/reactions/impl/InMemoryReactions.kt
package sol.vortexx.android.reactions.impl

import sol.vortexx.android.reactions.api.Reaction
import sol.vortexx.android.reactions.api.Reactions
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class InMemoryReactions @Inject constructor() : Reactions {
    private val counts = ConcurrentHashMap<Long, MutableMap<String, Int>>()
    private val mine = ConcurrentHashMap<Long, MutableSet<String>>()

    override fun reactionsFor(messageId: Long): List<Reaction> {
        val c = counts[messageId] ?: return emptyList()
        val m = mine[messageId] ?: emptySet()
        return c.map { (emoji, n) -> Reaction(messageId, emoji, n, emoji in m) }
            .sortedByDescending { it.count }
    }

    @Synchronized
    override fun apply(messageId: Long, emoji: String, delta: Int, byMe: Boolean) {
        val c = counts.getOrPut(messageId) { mutableMapOf() }
        val next = (c[emoji] ?: 0) + delta
        if (next <= 0) c.remove(emoji) else c[emoji] = next
        if (byMe) {
            val m = mine.getOrPut(messageId) { mutableSetOf() }
            if (delta > 0) m.add(emoji) else m.remove(emoji)
        }
    }
}
